### PR TITLE
[CHK-3168] Adding data details on trninfo

### PR DESCRIPTION
**Pull request by a-mannia**
#### List of Changes

- Add NPG client configuration
- Add a call to getOrder to retrieve gateway info and store it in DeadletterTransactionInfo.

#### Motivation and Context

We need to enhance the information in the dead letter document. 
This PR populates the DeadLetterTransactionInfoDetails field by making a call to the gateway, either of type NPG or Redirect, which handled the calls.
If the call to the NPG gateway fails, the process doesn't raise an exception; instead, an empty object is inserted into the dead letter.

#### How Has This Been Tested?
Unit Test

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

depends on https://github.com/pagopa/pagopa-ecommerce-commons/pull/217
depends on https://github.com/pagopa/pagopa-ecommerce-transactions-scheduler-service/pull/139

### CODE REVIEW

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/services/TransactionInfoService.kt [hunk 1]
```diff
@@ -35,90 +50,250 @@ class TransactionInfoService(
...
+                                        "Invalid operation retrived from gateway for transaction: ${baseTransaction.transactionId.value()}"
+                                    )
+                                )
+                        }
+                    }
+            TransactionAuthorizationRequestData.PaymentGateway.REDIRECT ->
+                Mono.just(DeadLetterRedirectTransactionInfoDetailsData(""))
+            else ->
+                Mono.error(
+                    NullPointerException(
```
Comment (scaminati-bv) [inline src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/services/TransactionInfoService.kt:239]: It would be better to use custom exception

**Comment by EmanueleBVtech** [comment]
i was thinking we can try to reduce lines of code of getTransactionInfoDetails,
here is a basic suggestion

```
fun getTransactionInfoDetails(baseTransaction: BaseTransaction): Mono<DeadLetterTransactionInfoDetailsData> {
    val transactionActivatedData = getTransactionActivatedData(baseTransaction)
    val transactionAuthorizationRequestData = getTransactionAuthRequestedData(baseTransaction)
    val correlationId = (transactionActivatedData?.transactionGatewayActivationData as NpgTransactionGatewayActivationData).correlationId

    CommonLogger.logger.info(
        "Retrieve gateway info for transactionId: [{}], gateway: [{}]",
        baseTransaction.transactionId,
        transactionAuthorizationRequestData?.paymentGateway
    )

    return when (transactionAuthorizationRequestData?.paymentGateway) {
        TransactionAuthorizationRequestData.PaymentGateway.NPG -> performGetOrderNPG(
            transactionId = TransactionId(baseTransaction.transactionId.value()),
            orderId = transactionAuthorizationRequestData.authorizationRequestId,
            pspId = transactionAuthorizationRequestData.pspId,
            correlationId = correlationId,
            paymentMethod = PaymentMethod.valueOf(transactionAuthorizationRequestData.paymentMethodName)
        ).doOnNext { order ->
            CommonLogger.logger.info(
                "Performed get order for transaction with id: [{}], last operation result: [{}], operations: [{}]",
                baseTransaction.transactionId,
                order.orderStatus?.lastOperationType,
                order.operations?.joinToString { "${it.operationType}-${it.operationResult}" }
            )
        }.flatMap { orderResponse ->
            orderResponse.operations?.fold(
                UnknownNpgOrderStatus(orderResponse) as NpgOrderStatus,
                this::reduceOperations
            )?.toMono()?.map {
                if (it is NpgOrderRefunded) {
                    it.copy(authorization = orderResponse.operations?.find(IS_AUTHORIZED))
                } else {
                    it
                }
            } ?: Mono.error(RuntimeException("No operations found for transaction with id: ${baseTransaction.transactionId.value()}"))
        }.handle { it, sink ->
            when (it) {
                is NgpOrderAuthorized -> sink.next(DeadLetterNpgTransactionInfoDetailsData(it.authorization.operationResult, it.authorization.operationId, correlationId))
                is NgpOrderNotAuthorized -> sink.next(DeadLetterNpgTransactionInfoDetailsData(it.operation.operationResult, it.operation.operationId, correlationId))
                is NpgOrderRefunded -> sink.next(DeadLetterNpgTransactionInfoDetailsData(it.refundOperation.operationResult, it.refundOperation.operationId, correlationId))
                else -> sink.error(RuntimeException("Invalid operation retrieved from gateway for transaction: ${baseTransaction.transactionId.value()}"))
            }
        }
        TransactionAuthorizationRequestData.PaymentGateway.REDIRECT -> Mono.just(DeadLetterRedirectTransactionInfoDetailsData(""))
        else -> Mono.error(NullPointerException("Transaction with id: ${baseTransaction.transactionId} has invalid gateway"))
    }
}
```

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/services/TransactionInfoService.kt
Comment (EmanueleBVtech) [inline src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/services/TransactionInfoService.kt:1]: i was thinking we can try to reduce lines of code of getTransactionInfoDetails,
here is a basic suggestion

```
fun getTransactionInfoDetails(baseTransaction: BaseTransaction): Mono<DeadLetterTransactionInfoDetailsData> {
    val transactionActivatedData = getTransactionActivatedData(baseTransaction)
    val transactionAuthorizationRequestData = getTransactionAuthRequestedData(baseTransaction)
    val correlationId = (transactionActivatedData?.transactionGatewayActivationData as NpgTransactionGatewayActivationData).correlationId

    CommonLogger.logger.info(
        "Retrieve gateway info for transactionId: [{}], gateway: [{}]",
        baseTransaction.transactionId,
        transactionAuthorizationRequestData?.paymentGateway
    )

    return when (transactionAuthorizationRequestData?.paymentGateway) {
        TransactionAuthorizationRequestData.PaymentGateway.NPG -> performGetOrderNPG(
            transactionId = TransactionId(baseTransaction.transactionId.value()),
            orderId = transactionAuthorizationRequestData.authorizationRequestId,
            pspId = transactionAuthorizationRequestData.pspId,
            correlationId = correlationId,
            paymentMethod = PaymentMethod.valueOf(transactionAuthorizationRequestData.paymentMethodName)
        ).doOnNext { order ->
            CommonLogger.logger.info(
                "Performed get order for transaction with id: [{}], last operation result: [{}], operations: [{}]",
                baseTransaction.transactionId,
                order.orderStatus?.lastOperationType,
                order.operations?.joinToString { "${it.operationType}-${it.operationResult}" }
            )
        }.flatMap { orderResponse ->
            orderResponse.operations?.fold(
                UnknownNpgOrderStatus(orderResponse) as NpgOrderStatus,
                this::reduceOperations
            )?.toMono()?.map {
                if (it is NpgOrderRefunded) {
                    it.copy(authorization = orderResponse.operations?.find(IS_AUTHORIZED))
                } else {
                    it
                }
            } ?: Mono.error(RuntimeException("No operations found for transaction with id: ${baseTransaction.transactionId.value()}"))
        }.handle { it, sink ->
            when (it) {
                is NgpOrderAuthorized -> sink.next(DeadLetterNpgTransactionInfoDetailsData(it.authorization.operationResult, it.authorization.operationId, correlationId))
                is NgpOrderNotAuthorized -> sink.next(DeadLetterNpgTransactionInfoDetailsData(it.operation.operationResult, it.operation.operationId, correlationId))
                is NpgOrderRefunded -> sink.next(DeadLetterNpgTransactionInfoDetailsData(it.refundOperation.operationResult, it.refundOperation.operationId, correlationId))
                else -> sink.error(RuntimeException("Invalid operation retrieved from gateway for transaction: ${baseTransaction.transactionId.value()}"))
            }
        }
        TransactionAuthorizationRequestData.PaymentGateway.REDIRECT -> Mono.just(DeadLetterRedirectTransactionInfoDetailsData(""))
        else -> Mono.error(NullPointerException("Transaction with id: ${baseTransaction.transactionId} has invalid gateway"))
    }
}
```

File: src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/services/TransactionInfoService.kt (hunk 1 above)
Comment (a-mannia) [reply src/main/kotlin/it/pagopa/ecommerce/transactions/scheduler/services/TransactionInfoService.kt]: done here: e9b655bb

File: helm/values-prod.yaml [hunk 2]
```diff
@@ -53,11 +53,28 @@ microservice-chart:
       DEAD_LETTER_LISTENER_NOTIFICATION_QUEUE_NAME: "pagopa-p-weu-ecommerce-notifications-service-errors-queue"
       ECOMMERCE_STORAGE_DEAD_LETTER_QUEUE_ACCOUNT_NAME: pagopapweuecommercedlsa
       ECOMMERCE_STORAGE_DEAD_LETTER_QUEUE_ENDPOINT: https://pagopapweuecommercedlsa.queue.core.windows.net/pagopapweuecommercedlsa
+      NPG_URI: https://xpay.nexigroup.com/api/phoenix-0.0
+      NPG_READ_TIMEOUT: "10000"
+      NPG_CONNECTION_TIMEOUT: "10000"
+      NPG_CARDS_PSP_LIST: "BCITITMM,CIPBITMM,BIC36019,UNCRITMM,BPPIITRRXXX,PPAYITR1XXX,BNLIITRR,PASCITMM,WOLLNLB1,BPMOIT22,SARDIT31,IFSPIT21"
+      NPG_PAYPAL_PSP_LIST: "" #TODO to be filled with psp id list that are expected to be present into psp api keys configuration
```
Comment (infantesimone) [inline helm/values-prod.yaml:60]: please add list as in UAT

Comment (a-mannia) [reply helm/values-prod.yaml]: Done here: 12f2abaf

**Pull request merged**